### ISSUE TITLE
Need help with tangents

### ISSUE THREAD

**Issue by tigrazone** [gameknife/gkNextRenderer#101]
Please help me with tangents implementation.
I do like [Sacha Willems do](https://github.com/SaschaWillems/VulkanPathTracer)
but renderer works wrong

Take a look at [my fork](https://github.com/tigrazone/gkNextRenderer)

**Comment by gameknife** [comment]
https://github.com/gameknife/gkNextRenderer/commit/df28dcc7a6c830813ee80ee9b6d69a241f39a99b#diff-141af648cdb8a0f3669805cd950303feebf0f4d872fe351b41f81a5dc0d4ed49R38

here you type v2 to a repeat v1?

**Comment by tigrazone** [comment]
I fix typo and do some additions to code but works not always ok.
Any ideas how to solve it?

**Comment by tigrazone** [comment]
Maybe you have your own method to work with tangent and can show me your approach

**Comment by gameknife** [comment]
> Maybe you have your own method to work with tangent and can show me your approach

tangent space is a little weired, I had a plan to impl normalmap, but didn;t start it so quick. I will add to recent todolist.

**Comment by tigrazone** [comment]
Help! Help!

**Comment by gameknife** [comment]
I'm working on a realtime game demo this week, normal map plan maybe the next, be patient~

**Comment by gameknife** [comment]
complete with https://github.com/gameknife/gkNextRenderer/commit/3aebc022db255e4ef850bd382fd0021c72e1266f
I add a basketball.glb to verify,
and I found the colorspace from glb have some problem, next i will fix it.

![PixPin_2024-10-30_00-14-48](https://github.com/user-attachments/assets/c505bae1-cf5c-4217-82c9-56ff14db3d88)

**Comment by tigrazone** [comment]
are you generate tangents with mikktspace?
why not read tangents from gltf?

**Comment by gameknife** [comment]
> are you generate tangents with mikktspace? why not read tangents from gltf?

not all gltf contain tangent, and mikktspace is a general method between all dcc tools. regeneration is not bad choice.

**Comment by tigrazone** [comment]
Yes, not bad. But if tangents exists, use it. Else - fix/regenerate

**Comment by gameknife** [comment]
Accepted, I will modify the gltf loader, use TANGENT attrib if gltf contained.

**Comment by tigrazone** [comment]
[Nvidia's vk_gltf_renderer ](https://github.com/nvpro-samples/vk_gltf_renderer) get tangent info from TANGENT attribute and _if needed_ on demand - fix/regenerate tangents.

One more thing - use these t, b, n in ggx calculations, not recalculate

**Comment by tigrazone** [comment]
[we have talk about non-trivial issue fixes](https://github.com/nvpro-samples/vk_gltf_renderer/issues/19)

**Comment by gameknife** [comment]
> [Nvidia's vk_gltf_renderer ](https://github.com/nvpro-samples/vk_gltf_renderer) get tangent info from TANGENT attribute and _if needed_ on demand - fix/regenerate tangents.
> 
> One more thing - use these t, b, n in ggx calculations, not recalculate

did you mean the ggxsampling should calculate in the mesh's tangent space, instead of calculated ONB space?

**Comment by tigrazone** [comment]
ggxsampling must to use tangents and normals from normal mapping, TANGENTS attributes. 
bitangent always calculated

**Comment by tigrazone** [comment]
tangents, bitangents also used in anisotropy - in big renderers

**Comment by tigrazone** [comment]
Are you here?
I need response and finish work with tangents

**Comment by gameknife** [comment]
> ggxsampling must to use tangents and normals from normal mapping, TANGENTS attributes. bitangent always calculated

it seems not rightway to use the hit surface's tangent space to do the ggxsampling

currently, it use the reflected ray dir to build the own ONB space then ggxsampling, the result is same as Blender RenderOutput

when I changed the ONB space to the hit surface tangent space, the lighting result has been offsetted.

could you show me how the big renderers doing?

**Comment by tigrazone** [comment]
yes. [Nvidia's vk_gltf_renderer](https://github.com/nvpro-samples/vk_gltf_renderer) use tangents from gltf or recreate by mikktspace or own method

**Comment by tigrazone** [comment]
also I use it in [my not fresh fork of your renderer](https://github.com/tigrazone/gkNextRenderer-my-private-forward)
I add you to collaborators of this private repo. 
Take a look

**Comment by gameknife** [comment]
I take a lookat your TBN, you rebuild a TBN based on the normalmap sampled normal, so the tangent & bitangent will not as same as the original one (which grow through the UV), is that a rightway?

**Comment by tigrazone** [comment]
I use your code of load and recreate tangents.
And I use it in shader. Shader code partially from vk_gltf_renderer
